internal/scheduler/nodedb/nodeiteration.go

```diff
@@ -0,0 +1,413 @@
+package nodedb
+
+import (
+	"bytes"
+	"container/heap"
+	"encoding/binary"
+	"fmt"
+
+	"github.com/hashicorp/go-memdb"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+
+	"github.com/armadaproject/armada/internal/scheduler/schedulerobjects"
+)
+
+const NodeDominantQueueWildcard = "*"
+
+type NodeIterator interface {
+	NextNode() *schedulerobjects.Node
+}
+
+// NodesIterator is an iterator over all nodes in the db.
+type NodesIterator struct {
+	it memdb.ResultIterator
+}
+
+func NewNodesIterator(txn *memdb.Txn) (*NodesIterator, error) {
+	it, err := txn.LowerBound("nodes", "id", "")
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	return &NodesIterator{
+		it: it,
+	}, nil
+}
+
+func (it *NodesIterator) WatchCh() <-chan struct{} {
+	panic("not implemented")
+}
+
+func (it *NodesIterator) NextNode() *schedulerobjects.Node {
+	obj := it.it.Next()
+	if obj == nil {
+		return nil
+	}
+	node, ok := obj.(*schedulerobjects.Node)
+	if !ok {
+		panic(fmt.Sprintf("expected *Node, but got %T", obj))
+	}
+	return node
+}
+
+func (it *NodesIterator) Next() interface{} {
+	return it.NextNode()
+}
+
+type NodePairIterator struct {
+	itA   *NodesIterator
+	itB   *NodesIterator
+	nodeA *schedulerobjects.Node
+	nodeB *schedulerobjects.Node
+}
+
+type NodePairIteratorItem struct {
+	NodeA *schedulerobjects.Node
+	NodeB *schedulerobjects.Node
+}
+
+func NewNodePairIterator(txnA, txnB *memdb.Txn) (*NodePairIterator, error) {
+	itA, err := NewNodesIterator(txnA)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	itB, err := NewNodesIterator(txnB)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	return &NodePairIterator{
+		itA: itA,
+		itB: itB,
+	}, nil
+}
+
+func (it *NodePairIterator) WatchCh() <-chan struct{} {
+	panic("not implemented")
+}
+
+func (it *NodePairIterator) NextItem() (rv *NodePairIteratorItem) {
+	defer func() {
+		if rv == nil {
+			return
+		}
+		if rv.NodeA != nil {
```

Comment:
I'm not sure I like the use of a deferred function here to  set these values to nil, nor really the use of nil to indicate that there might be more data available.  The general convention is that if a value is nil, it indicates that no more data is available.

Reply:
I interpret `nil` for either `A` or `B` here to mean that there's no entry for the